File: VulkanGameEngine/GameObjectComponent.cpp
#include "GameObjectComponent.h"

GameObjectComponent::GameObjectComponent()
{

}

GameObjectComponent::GameObjectComponent(uint32 gameObjectId, ComponentTypeEnum componentType)
{
    Name = "default";
    ParentGameObjectID = gameObjectId;
    ComponentType = componentType;
}

GameObjectComponent::GameObjectComponent(uint32 gameObjectId, const String& name, ComponentTypeEnum componentType)
{
    Name = name;
    ParentGameObjectID = gameObjectId;
    ComponentType = componentType;

}

GameObjectComponent::~GameObjectComponent()
{
}

void GameObjectComponent::Input(const float& deltaTime)
{
    //CSobject->InvokeMethod("Input", KeyboardKeyCode::KEY_D, deltaTime);
}

void GameObjectComponent::Update(VkCommandBuffer& commandBuffer, const float& deltaTime)
{
    //CSobject->InvokeMethod("BufferUpdate", commandBuffer, deltaTime);
}

void GameObjectComponent::Draw(VkCommandBuffer& commandBuffer, VkPipeline& pipeline, VkPipelineLayout& pipelineLayout, Vector<VkDescriptorSet>& descriptorSetList)
{

}

void GameObjectComponent::Destroy()
{
    //CSobject->InvokeMethod("Destroy");
}

SharedPtr<GameObjectComponent> GameObjectComponent::Clone() const
{
    return SharedPtr<GameObjectComponent>();
}

size_t GameObjectComponent::GetMemorySize() const
{
    return sizeof(GameObjectComponent);
}

File: VulkanGameEngineLevelEditor/Components/GameObjectComponent.cs
using Newtonsoft.Json;
using VulkanGameEngineLevelEditor.GameEngineAPI;
using VulkanGameEngineLevelEditor.Models;
using VulkanGameEngineLevelEditor.Vulkan;

namespace VulkanGameEngineLevelEditor.Components
{
    public unsafe class GameObjectComponent
    {
        [JsonIgnore]
        public GameObject ParentGameObject { get; set; }

        public string Name { get; set; }
        public ComponentTypeEnum ComponentType { get; set; }

        public GameObjectComponent()
        {

        }

        public GameObjectComponent(GameObject parentGameObject, ComponentTypeEnum componentType)
        {
            ParentGameObject = parentGameObject;
            Name = "GameObjectComponent";
            ComponentType = componentType;
        }

        public GameObjectComponent(GameObject parentGameObject, string name, ComponentTypeEnum componentType)
        {
            ParentGameObject = parentGameObject;
            Name = name;
            ComponentType = componentType;
        }

        public GameObjectComponent(GameObject parentGameObject, GameObjectComponentModel model)
        {
            ParentGameObject = parentGameObject;
            Name = model.Name;
            ComponentType = model.ComponentType;
        }

        public virtual void Input(KeyBoardKeys key, float deltaTime)
        {
        }

        public virtual void Update(VkCommandBuffer commandBuffer, float deltaTime)
        {

        }

        public virtual void Draw(VkCommandBuffer commandBuffer, VkPipeline pipeline, VkPipelineLayout pipelineLayout, ListPtr<VkDescriptorSet> descriptorSetList, SceneDataBuffer sceneProperties)
        {

        }

        public virtual void Destroy()
        {

        }

        public virtual int GetMemorySize()
        {
            return sizeof(GameObjectComponent);
        }

        //public IntPtr GetCPPgameObjectPtr()
        //{
        //    return CPPgameObjectPtr;
        //}

        //public IntPtr GetCPPComponentPtr()
        //{
        //    return CPPcomponentPtr;
        //}

        //public NativeString* GetPositionPtr()
        //{
        //    fixed (NativeString* positionPointer = &Name)
        //    {
        //        return positionPointer;
        //    }
        //}

        //public unsafe ComponentTypeEnum* GetComponentTypePtr()
        //{
        //    // Get the address of the field: `ComponentType`
        //    fixed (ComponentTypeEnum* ptr = &ComponentType)
        //    {
        //        return ptr;
        //    }
        //}
    }
}
